### ISSUE TITLE
CJK brackets at the beginning of paragraph

### ISSUE THREAD

**Issue by minamotorin** [typst/typst#4011]
### Description

When paragraph indentation is enabled, positioning of brackets at the beginning of paragraph doesn't follow [§3.1.5 of Requirements for Japanese Text Layout](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#positioning_of_opening_brackets_at_line_head).
Example code is below.

```typst
// This program is public domain (CC0).
#set page(
  width: 210mm,
  margin: (left: 30mm, right: 210mm-30mm-10em),
)
#set par(
  leading: 1em
)
#show par: set block(
  spacing: 2.5em
)

「日日日日日日日日日「日

#set par(first-line-indent: 1em)

「日日日日日日日日「日

#h(-0.5em)「日日日日日日日日「日
```

![brac](https://github.com/typst/typst/assets/76122224/8c71aede-8f58-410c-bffe-8893b8c31209)

The first paragraph is an example of paragraph indentation is **not** enabled, and the second one is an example of paragraph indentation is enabled.

The third paragraph is an example of the expected result that follows [§3.1.5 of Requirements for Japanese Text Layout](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#positioning_of_opening_brackets_at_line_head).

Thanks.

### Operating system

macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by Leedehai** [comment]
https://typst.app/docs/reference/model/par/#parameters-first-line-indent mentioned "Only the first line of a consecutive paragraph will be indented (not the first one in a block or on the page)." -- a limitation, yes.. Typst wants to change that but it's not trivial.

**Comment by laurmaedje** [comment]
I think the issue here differs from #311. If I'm understanding correctly, it's more that a bracket shouldn't disrupt the grid-like alignment of characters.

**Comment by minamotorin** [comment]
The key point of this issue is that the space between left edge of paragraph and “「” is effectively 1.5 em, not the 1 em specified by `#set par(first-line-indent: 1em)`.

**Comment by peng1999** [comment]
> The third paragraph is an example of the expected result that follows [§3.1.5 of Requirements for Japanese Text Layout](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#positioning_of_opening_brackets_at_line_head).

Note that JLREQ provides three possible options in that chapter, and Typst just behaves the same as option ③ in [Figure 71](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#fig2_15). So the current behavior is indeed conform to JLREQ. This issue should be a feature request rather than a bug report. 

cc @Enivex maybe you want to change the issue tags.

**Comment by minamotorin** [comment]
@peng1999 
> Typst just behaves the same as option ③ in [Figure 71](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#fig2_15).

No. Option 3 require the space between the left edge of the paragraph and “「” **at the beginning of a paragraph** to be 1/2 em, but  the actual space is 3/2 em as I described first.
This is same as option 2 rather than option 3.

However, option 2 require the space between the left edge of the paragraph and “「” **at the beginning of a line in the middle of a paragraph** to be 1/2 em. But the actual space is 0 em.
This is same as option 1 or option 3.

> ![Figure 71 of JLREQ](https://github.com/typst/typst/assets/76122224/6d0b9e7a-9ba1-4e96-9df1-fc244628d36e)
> (Quoted from [Figure 71 of Requirements for Japanese Text Layout](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#fig2_15) by W3C, [permissive document license](https://www.w3.org/Consortium/Legal/2015/copyright-software-and-document).

---

> feature request

Of course, it is better the user can choose which option to use.

**Comment by peng1999** [comment]
> No. Option 3 require the space between the left edge of the paragraph and “「” **at the beginning of a paragraph** to be 1/2 em, but the actual space is 3/2 em as I described first.

Sorry, I made a mistake. The problem you provide do exist. And I made another mistake in previous comment: Typst currently follows option ①, not ③.

And the bug is: CJK punctuation at the start of paragraphs are not adjusted sometimes, and it is tracked in #2348.

So your issue splits into two parts: the bug part is a duplication of #2348, and the feature request part is valid. But I suggest change the name the issue to something like "Options for Positioning of Opening Brackets".

**Comment by minamotorin** [comment]
@peng1999 

> So your issue splits into two parts: the bug part is a duplication of https://github.com/typst/typst/issues/2348, and the feature request part is valid.

I couldn't find #2348 in the past and this issue is duplicate of it. Thanks!

I close this because this is just duplicate of #2348 and I think that feature request should be discussed in separated new issue.

**Comment by minamotorin** [comment]
This is not duplicate of #2348.

The cause of #2348 seems to be that the space between the left edge of the paragraph and “「” at the beginning of a paragraph is used to adjust the line length.
But this issue is about the amount of the space.

**Comment by minamotorin** [comment]
@peng1999 
I think current behavior of typst does not satisfy any of option in [Requirements for Japanese Text Layout](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#positioning_of_opening_brackets_at_line_head).
Option 1 require the space between the left edge of the paragraph and “「” at the beginning of a paragraph to be 1 em, but the actual space is 3/2 em.

Here is table about amount of indentation.

||paragraph starting with “「”|line starting with “「” and not beginning of  paragraph|normal beginning of paragraph|
|:-|-:|-:|-:|
|Option 1 of [§3.1.5 of Requirements for Japanese Text Layout](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#positioning_of_opening_brackets_at_line_head)|1 em|0 em|1 em|
|Option 2 of [§3.1.5 of Requirements for Japanese Text Layout](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#positioning_of_opening_brackets_at_line_head)|3/2 em|1/2 em|1 em|
|Option 3 of [§3.1.5 of Requirements for Japanese Text Layout](https://www.w3.org/TR/2020/NOTE-jlreq-20200811/#positioning_of_opening_brackets_at_line_head)|1/2 em|0 em|1 em|
|typst (with `#set par(first-line-indent: 1em)`)|3/2 em|0 em|1 em|

But if this is the typst specification, this is definitely a feature request.

**Comment by peng1999** [comment]
> The cause of #2348 seems to be that the space between the left edge of the paragraph and “「” at the beginning of a paragraph is used to adjust the line length.
> But this issue is about the amount of the space before it is used for adjustment.

Suppose we have `first-line-indent` set to 1em and the first character is `「`. Currently in Typst it becomes: <1em fixed space> + <0.5em shrinkable space> + <half width `「`>

- In #2348: The `「` is not adjusted sometimes = the shrinkable space is wrongly present sometimes. The solution is to remove it.
- In this issue: The problem is that when the shrinkable space presents, the space becomes 1.5em instead of 1em. The solution is to remove the shrinkable space.

In summary, the two issue is essentially the same and can be fixed by a single action: remove the shrinkable space.

**Comment by minamotorin** [comment]
> the two issue is essentially the same and can be fixed by a single action: remove the shrinkable space.

I agree about the solution.
But I'll keep this issue open because #2348 can be fixed with a different solution, make the space unshrinkable.
